### PR TITLE
feat(github): forkOrgs, forkCreate

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Adds new options forkOrgs and forkCreate

## Context

These features will be used by me in the Forking Renovate app a part of some necessary migrations.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
My goals are to:
- migrate the forking renovate app from using github.com/renovate-bot for all its forks and instead to an open source org (forkOrgs=[new, existing])
- use dual app tokens for both original and forked repos instead of a PAT, and therefore skip actual forking within the Renovate logic itself (forkCreate=false)

**Comment by viceice** [comment]
sounds good

**Comment by viceice** [comment]
btw: if you fork to an org, the original repo maintainer can no longer update renovate branches in the fork by default. you would need to add them with write access to the fork to get it working.

**Comment by rarkins** [comment]
> btw: if you fork to an org, the original repo maintainer can no longer update renovate branches in the fork by default. you would need to add them with write access to the fork to get it working.

The challenge is:
- I need to use app tokens instead of PAT
- App token with permissions to create a PR in the parent repo wouldn't have permissions to enable "maintainers can modify" in the forked repo anyway

I'll still have a PAT on the backend (for doing the forking out of band) so maybe I could add a scheduled background process which detects maintainer in the parent repo and invites them to the forked repos. They may need to accept that invite though (which isn't necessary for maintainers can modify).

https://github.com/orgs/community/discussions/5634

File: docs/usage/self-hosted-configuration.md [hunk 1]
```diff
@@ -376,9 +376,18 @@ In practice, it is implemented by converting the `force` configuration into a `p
...
+
+Set this to false if using an app token which is not permitted to fork.
+
+## forkOrgs
 
-You probably don't need this option - it is an experimental setting developed for the Forking Renovate hosted GitHub App.
+Configuring this value lets you choose which org/account you want repos forked into when fork mode is enabled.
+It also allows you to configure more than one if you are migrating from user-based forks into org-based ones.
+
+Renovate will first look if a fork exists in the preferred org, and then if not then look to see if one exists in any of the fallback orgs.
```
Comment (viceice) [inline docs/usage/self-hosted-configuration.md:388]: maybe add one more sentence, that renovate will create new fork in first org of the list?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: docs/usage/self-hosted-configuration.md (hunk 1 above)
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:388]:
```suggestion
This configuration option lets you choose which organization or account you want repositories forked into when "fork mode" is enabled.

You may set more than one organization or account.
This can be handy if you're migrating from user-based forks to organization-based forks.

If you've set multiple `forkOrgs` then Renovate will:

1. Check if a fork exists in the preferred organization
1. If no fork exists in the preferred org: Renovate checks the fallback organizations

Renovate always creates the new fork in the _first_ organization in the `forkOrgs` list.
```

How about this wording and layout?

> maybe add one more sentence, that renovate will create new fork in first org of the list?

My suggestion adds this sentence. 😉

File: lib/config/options/index.ts [hunk 2]
```diff
@@ -417,6 +417,27 @@ const options: RenovateOptions[] = [
     supportedPlatforms: ['github'],
     experimental: true,
   },
+  {
+    name: 'forkOrgs',
+    description:
+      'The preferred orgs to create or find forked repos, when in fork mode',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:423]:
```suggestion
      'The preferred organizations to create or find forked repositories, when in fork mode.',
```

File: lib/config/options/index.ts [hunk 3]
```diff
@@ -417,6 +417,27 @@ const options: RenovateOptions[] = [
...
+    stage: 'repository',
+    type: 'array',
+    subType: 'string',
+    globalOnly: true,
+    supportedPlatforms: ['github'],
+    experimental: true,
+  },
+  {
+    name: 'forkCreate',
+    description: 'Whether to create a fork at runtime when in fork mode',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:433]:
```suggestion
    description: 'Decide if Renovate creates a fork at runtime when in fork mode.',
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Comment by viceice** [comment]
@zharinov can we detect a comment to force a test run? so we don't need to move out of draft mode?

**Comment by zharinov** [comment]
I don't think there is an easy way to do it. Seems like you either have permissions to re-run jobs, or you're not.

**Review by viceice — COMMENTED** [review]
just a docs question, otherwise LGTM

File: docs/usage/self-hosted-configuration.md [hunk 4]
```diff
@@ -415,9 +415,25 @@ In practice, it is implemented by converting the `force` configuration into a `p
...
+You may set more than one organization or account.
+This can be handy if you're migrating from user-based forks to organization-based forks.
 
-You probably don't need this option - it is an experimental setting developed for the Forking Renovate hosted GitHub App.
+If you've set multiple `forkOrgs` then Renovate will:
+
+1. Check if a fork exists in the preferred organization
+1. If no fork exists in the preferred org: Renovate checks the fallback organizations
+
+Renovate always creates the new fork in the _first_ organization in the `forkOrgs` list.
```
[the remark this answers is not in the document]
Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:434]: I preferred this order because Renovate first checks for existence across all orgs and only if not found then does it create

**Review by viceice — APPROVED** [review]

**Review by JamieMagee — APPROVED** [review]
